core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -506,8 +537,25 @@ public Attributes getAttributes() {
     return Attributes.EMPTY;
   }
 
-  private void closeObserver(Listener<RespT> observer, Status status, Metadata trailers) {
-    observer.onClose(status, trailers);
+  private class DeadlineNotifyApplicationTimer implements Runnable {
+    private final long remainingNanos;
+
+    DeadlineNotifyApplicationTimer(long remainingNanos) {
+      this.remainingNanos = remainingNanos;
+    }
+
+    @Override
+    public void run() {
+      InsightBuilder insight = new InsightBuilder();
+      stream.appendTimeoutInsight(insight);
+      deadlineCancellationSendToServerFuture =
+          startDeadlineSendCancelToServerTimer(remainingNanos, insight);
+      // DelayedStream.cancel() is safe to call from a thread that is different from where the
+      // stream is created.
+      closeObserver(observer, DEADLINE_EXCEEDED.augmentDescription(
```

Comment:
`closeObserver()` must be called from `callExecutor` and inside a `ContextRunnable`. You need something similar to `class StreamClosed()` in this file.